compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDestructuringDeclarationChecker.kt

```diff
@@ -75,39 +73,69 @@ object FirDestructuringDeclarationChecker : FirPropertyChecker() {
                 else -> null
             } ?: return
 
-        when (val diagnostic = reference.diagnostic) {
-            is ConeUnresolvedNameError -> {
-                reporter.report(
-                    FirErrors.COMPONENT_FUNCTION_MISSING.on(
-                        originalDestructuringDeclarationOrInitializerSource,
-                        diagnostic.name,
-                        originalDestructuringDeclarationType
-                    ),
-                    context
-                )
-            }
-            is ConeAmbiguityError -> {
-                reporter.report(
-                    FirErrors.COMPONENT_FUNCTION_AMBIGUITY.on(
-                        originalDestructuringDeclarationOrInitializerSource,
-                        diagnostic.name,
-                        diagnostic.candidates
-                    ),
-                    context
-                )
-            }
-            is ConeInapplicableCandidateError -> {
-                if (originalDestructuringDeclarationType.isNullable) {
-                    reporter.report(
-                        FirErrors.COMPONENT_FUNCTION_ON_NULLABLE.on(
-                            originalDestructuringDeclarationOrInitializerSource,
-                            (diagnostic.candidate.symbol.fir as FirSimpleFunction).name
-                        ),
+        when (val reference = componentCall.calleeReference) {
+            is FirResolvedNamedReference -> {
+                val destructuringCall = (reference.resolvedSymbol.fir as? FirSimpleFunction) ?: return
+                val destructuringType = destructuringCall.returnTypeRef.coneType
+                if (destructuringType is ConeKotlinErrorType) {
+                    // There will be other errors on this error type.
+                    return
+                }
+                if (destructuringType is ConeTypeParameterType) {
+                    // As long as all involved constraints meet, this type parameter shouldn't be type mismatch,
```

Comment:
Consider:
```kt
data class C<T>(val x: Int, val y: T)
fun <T, S> foo(c: C<T>) {
  val (x: Int, y: S) = c
}
```

FE1.0 has `COMPONENT_FUNCTION_RETURN_TYPE_MISMATCH` error on `c`, but I think this is the case inside this block (`destructuringType is ConeTypeParameterType`). So I think we still need to check the declaration and destructuring types.

Reply:
Thanks. It turns out `componentCall` already has a substituted type, so we don't need special handling for type parameter. Also, it looks like there isn't a test like that, thanks for the example!